test-case/check-alsabat.sh

```diff
@@ -50,39 +56,51 @@ then
 	exit 2
 fi
 
+pcmid_p=${pcm_p:-1}
+codec_mode=$(func_check_nocodec_mode)
+
 [[ ${OPT_VALUE_lst['s']} -eq 1 ]] && func_lib_start_log_collect
 
-function __upload_wav_file
-{
-    # upload the alsabat wav file
-    for file in /tmp/bat.wav.*
-    do
-	    size=`ls -l $file | awk '{print $5}'`
-	    if [[ $size -gt 0 ]]; then
-		    cp $file $LOG_ROOT/
-	    fi
-    done
-}
+func_pipeline_export "$tplg" "type:playback & id:$pcmid_p"
+
+# parser the parameters of the specified playback pipeline
+channel=$(func_pipeline_parse_value 0 ch_max)
+rate=$(func_pipeline_parse_value 0 rate)
+fmts=$(func_pipeline_parse_value 0 fmts)
 
 # check the PCMs before alsabat test
 dlogi "check the PCMs before alsabat test"
-[[ $(aplay -Dplug$pcm_p -d 1 /dev/zero -q) ]] && die "Failed to play on PCM: $pcm_p"
-[[ $(arecord -Dplug$pcm_c -d 1 /dev/null -q) ]] && die "Failed to capture on PCM: $pcm_c"
+[[ $(aplay -Dplug"$pcm_p" -d 1 /dev/zero -q) ]] && die "Failed to play on PCM: $pcm_p"
+[[ $(arecord -Dplug"$pcm_c" -d 1 /dev/null -q) ]] && die "Failed to capture on PCM: $pcm_c"
 
 # alsabat test
-# different PCMs may support different audio formats(like samplerate, channel-counting, etc.).
-# use plughw to do the audio format conversions. So we don't need to specify them for each PCM.
-dlogc "alsabat -Pplug$pcm_p --standalone -n $frames -F $frequency"
-alsabat -Pplug$pcm_p --standalone -n $frames -F $frequency &
-# playback may have low latency, add one second delay to aviod recording zero at beginning.
-sleep 1
-dlogc "alsabat -Cplug$pcm_c -F $frequency"
-alsabat -Cplug$pcm_c -F $frequency
-
-# upload failed wav file
-if [[ $? != 0 ]]; then
-	__upload_wav_file
-	exit 1
-fi
-
-exit 0
+for format in $fmts
+    do
+        if [ "$format" == "S24_LE" ]; then
+            dlogi "S24_LE is not supported, skip to test this format"
+            continue
+        fi
+
+        dlogc "alsabat -P $pcm_p --standalone -n $frames -F $frequency -c $channel -r $rate -f $format"
+        alsabat -P "$pcm_p" --standalone -n "$frames" -F "$frequency" -c "$channel" -r "$rate" -f "$format" & alsabat_P=$!
+        # playback may have low latency, add one second delay to aviod recording zero at beginning.
+        sleep 1
+
+        if [ "$codec_mode" != "nocodec" ]; then
+            # USB sound card only supports 1 channel S16_LE format.
+            format_c=S16_LE
+            channel_c=1
+        else
+            format_c=$format
+            channel_c=$channel
+        fi
+
+        dlogc "alsabat -C $pcm_c -F $frequency -f $format_c -c $channel_c -r $rate"
+        alsabat -C "$pcm_c" -F "$frequency" -f "$format_c" -c "$channel_c" -r "$rate" || {
+            find /tmp -maxdepth 1 -type f -name "bat.wav.*" -size +0 -exec cp {} "$LOG_ROOT/" \;
```

Comment:
You probably want :
```shell
alsabat ... || {
  find ... || true
  exit 1
}
```
... because of `set -e`. When `find` fails, it fails with an undocumented return value (!) which could cause serious issues like #385

Reply:
Good suggestion, I have added a new function in lib.sh to upload wav files.